File: app/src/main/java/info/nightscout/androidaps/plugins/general/smsCommunicator/AuthRequest.java
package info.nightscout.androidaps.plugins.general.smsCommunicator;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import info.nightscout.androidaps.Constants;
import info.nightscout.androidaps.R;
import info.nightscout.androidaps.logging.L;
import info.nightscout.androidaps.utils.DateUtil;

class AuthRequest {
    private static Logger log = LoggerFactory.getLogger(L.SMS);

    Sms requester;
    String confirmCode;
    private Runnable action;

    private long date;

    private boolean processed;
    private SmsCommunicatorPlugin plugin;

    AuthRequest(SmsCommunicatorPlugin plugin, Sms requester, String requestText, String confirmCode, SmsAction action) {
        this.requester = requester;
        this.confirmCode = confirmCode;
        this.action = action;
        this.plugin = plugin;

        this.date = DateUtil.now();

        plugin.sendSMS(new Sms(requester.phoneNumber, requestText));
    }

    void action(String codeReceived) {
        if (processed) {
            if (L.isEnabled(L.SMS))
                log.debug("Already processed");
            return;
        }
        if (!confirmCode.equals(codeReceived)) {
            processed = true;
            if (L.isEnabled(L.SMS))
                log.debug("Wrong code");
            plugin.sendSMS(new Sms(requester.phoneNumber, R.string.sms_wrongcode));
            return;
        }
        if (DateUtil.now() - date < Constants.SMS_CONFIRM_TIMEOUT) {
            processed = true;
            if (L.isEnabled(L.SMS))
                log.debug("Processing confirmed SMS: " + requester.text);
            if (action != null)
                action.run();
            return;
        }
        if (L.isEnabled(L.SMS))
            log.debug("Timed out SMS: " + requester.text);
    }

}

File: app/src/main/java/info/nightscout/androidaps/plugins/general/smsCommunicator/AuthRequest.kt
package info.nightscout.androidaps.plugins.general.smsCommunicator

import android.os.SystemClock
import dagger.android.HasAndroidInjector
import info.nightscout.androidaps.Constants
import info.nightscout.androidaps.R
import info.nightscout.androidaps.interfaces.CommandQueueProvider
import info.nightscout.androidaps.logging.AAPSLogger
import info.nightscout.androidaps.logging.LTag
import info.nightscout.androidaps.plugins.general.smsCommunicator.otp.OneTimePassword
import info.nightscout.androidaps.plugins.general.smsCommunicator.otp.OneTimePasswordValidationResult
import info.nightscout.androidaps.utils.DateUtil
import info.nightscout.androidaps.utils.T
import info.nightscout.androidaps.utils.resources.ResourceHelper
import javax.inject.Inject

class AuthRequest internal constructor(
    injector: HasAndroidInjector,
    var requester: Sms,
    requestText: String,
    var confirmCode: String,
    val action: SmsAction) {

    @Inject lateinit var aapsLogger: AAPSLogger
    @Inject lateinit var smsCommunicatorPlugin: SmsCommunicatorPlugin
    @Inject lateinit var resourceHelper: ResourceHelper
    @Inject lateinit var otp: OneTimePassword
    @Inject lateinit var dateUtil: DateUtil
    @Inject lateinit var commandQueue: CommandQueueProvider

    private var date = 0L
    private var processed = false

    init {
        injector.androidInjector().inject(this)
        date = dateUtil.now()
        smsCommunicatorPlugin.sendSMS(Sms(requester.phoneNumber, requestText))
    }

    private fun codeIsValid(toValidate: String): Boolean =
        otp.checkOTP(toValidate) == OneTimePasswordValidationResult.OK

    fun action(codeReceived: String) {
        if (processed) {
            aapsLogger.debug(LTag.SMS, "Already processed")
            return
        }
        if (!codeIsValid(codeReceived)) {
            processed = true
            aapsLogger.debug(LTag.SMS, "Wrong code")
            smsCommunicatorPlugin.sendSMS(Sms(requester.phoneNumber, resourceHelper.gs(R.string.sms_wrongcode)))
            return
        }
        if (dateUtil.now() - date < Constants.SMS_CONFIRM_TIMEOUT) {
            processed = true
            if (action.pumpCommand) {
                val start = dateUtil.now()
                //wait for empty queue
                while (start + T.mins(3).msecs() > dateUtil.now()) {
                    if (commandQueue.size() == 0) break
                    SystemClock.sleep(100)
                }
                if (commandQueue.size() != 0) {
                    aapsLogger.debug(LTag.SMS, "Command timed out: " + requester.text)
                    smsCommunicatorPlugin.sendSMS(Sms(requester.phoneNumber, resourceHelper.gs(R.string.sms_timeout_while_wating)))
                    return
                }
            }
            aapsLogger.debug(LTag.SMS, "Processing confirmed SMS: " + requester.text)
            action.run()
            return
        }
        aapsLogger.debug(LTag.SMS, "Timed out SMS: " + requester.text)
    }
}